syntax = "proto2";

package config;

// The config package contains only the class of ConfigParameter
message ConfigParameter {
  required string model_file = 1; // model prototxt file
  optional int32 tech_node = 2 [default = 28];  // technology node
  required int32 bit_width = 3;   // required bit width

  enum MemoryType {
    ROM = 0;
    RAM = 1;
  };

  // memory type
  optional MemoryType memory_type = 4 [default = ROM];

  // simulation related setting
  required int32 reset_period = 5;  // reset period clock cycle
  required int32 sim_period = 6;    // simulation period clock cycle
  optional double clk_freq = 7 [default = 1.];  // clock frequency [GHz]

  // append buffer size after each processing element
  optional int32 append_buffer_capacity = 8 [default = 0];

  // inference rate [cycle] (w.r.t. pixel)
  optional int32 pixel_inference_rate = 9 [default = 100];

  // tracked trace file name
  optional string trace_file = 10 [default = ""];

  // advanced settings: for debug issues
  optional int32 early_stop_frame_size = 11 [default = 1];
  optional bool kernel_unrolling_flag = 12 [default = false];
}
